### PR TITLE
O3DE engine Gem rename support

### LINKED ISSUE #17034: Renamed gem name in gem.json has linker errors in monolithic release builds
**Describe the bug**
If you copy a gem and rename it in gem.json (e.g. AudioEngineWwise -> AudioEngineWwiseExtra) you will get linker errors about a missing module when building in release mode.

The goal is to be able to include a gem in your project's folder but rename it so that it doesn't collide with any registered gems that have the same name.  This is useful when making minor edits to a gem, or including a gem in your project so users don't have to download it separately.

A workaround to this issue is to change the `version` field in `gem.json` and then specify in your `project.json` that you want your project to use that exact version using the format `<gem name>==<version>`

**Assets required**
An o3de project to modfiy

**Steps to reproduce**
1. Copy the `AudioEngineWwise` gem into your project path
2. Edit your project.json and add `AudioEngineWwise` as an external_subdir and rename `AudioEngineWwise` in `gem_names` to `AudioEngineWwiseExtra`
3. Edit the `gem.json` in `AudioEngineWwise` folder inside your project and change the name to `AudioEngineWwiseExtra`
4. configure and build a monolithic release game launcher
5. See error

**Expected behavior**
configure and build succeeds

**Actual behavior**
build fails with missing symbols error looking for the old gem module name

**Found in Branch**
dev

**Commit ID from [o3de/o3de](https://github.com/o3de/o3de) Repository**
735a539283600acb666292dc6415de2d26229189

**Desktop/Device (please complete the following information):**
 - Device: PC
 - OS: Windows
 - Version 10

### PR DESCRIPTION
**Pull request by lumberyard-employee-dm**
The following changes have been made to all of the Gems that come with the O3DE engine source code to make it easier to rename a Gem.

Updated all of the Gems CMakeLists.txt to use the "${gem_name}" variable when referencing the Gem in CMake Targets.

This allows the name of the CMake targets to update whenever the "gem_name" in the gem.json is changed.

Updated all of the `AZ_DECLARE_MODULE_CLASS` macro invocations to reference the `O3DE_GEM_NAME` preprocessor token which is use to substitute the "gem_name" in the gem.json for the extern "C" `CreateModuleClass_Gem_*` function which the launcher_generator.cmake invokes in Monolithic builds to create an AZ::Module derived class instance for each active gem.

resolves #17034

## How was this PR tested?

Built Monolithic and Non-Monolithic SDK as well as built the AutomatedTesting project in Monolithic and non-Monolithic

### CODE REVIEW

File: AutomatedTesting/Gem/CMakeLists.txt [hunk 1]
```diff
@@ -6,9 +6,30 @@
...
+# Read the version field from the gem.json
+set(gem_version, "0.0.0")
+o3de_read_json_key(gem_version ${gem_json} "version")
+
 o3de_restricted_path(${gem_json} gem_restricted_path gem_parent_relative_path)
 
+o3de_pal_dir(pal_dir ${CMAKE_CURRENT_SOURCE_DIR}/Platform/${PAL_PLATFORM_NAME} "${gem_restricted_path}" "${gem_path}" "${gem_parent_relative_path}")
+
+ly_add_external_target_path(${CMAKE_CURRENT_SOURCE_DIR}/3rdParty)
+
```
Comment (AMZN-alexpete) [inline AutomatedTesting/Gem/CMakeLists.txt:32]: What do you think about putting this common code into a macro in `Gems.cmake` and calling it with something like:
```
o3de_gem_common_setup("<fall back gem name>", "<fall back version")
```

File: Gems/AssetValidation/Code/CMakeLists.txt [hunk 2]
```diff
@@ -35,15 +35,24 @@ ly_add_target(
...
     BUILD_DEPENDENCIES
         PRIVATE
-            Gem::AssetValidation.Static
+            Gem::${gem_name}.Static
 )
 
+# Inject the gem name into the Module source file
+ly_add_source_properties(
+    SOURCES
+        Source/AssetValidationModule.cpp
```
Comment (AMZN-alexpete) [inline Gems/AssetValidation/Code/CMakeLists.txt:44]: Probably wouldn't work for all Gems but could possibly do
```suggestion
        Source/${gem_name}Module.cpp
```

File: Gems/Atom/Asset/Shader/Code/CMakeLists.txt [hunk 3]
```diff
@@ -92,24 +103,33 @@ ly_add_target(
...
+ly_add_source_properties(
+    SOURCES
+        Source/Editor/AzslShaderBuilderModule.cpp
+    PROPERTY COMPILE_DEFINITIONS
+        VALUES
+            O3DE_GEM_NAME=${gem_name}
+            O3DE_GEM_VERSION=${gem_version})
+
+# Create an alias to the old Atom_Asset_Shader.Builders target for backwards compatiblity
+ly_create_alias(NAME Atom_Asset_Shader.Builders NAMESPACE Gem TARGETS Gem::${gem_name}.Builders)
```
Comment (AMZN-alexpete) [inline Gems/Atom/Asset/Shader/Code/CMakeLists.txt:124]: good call

**Review by AMZN-alexpete — APPROVED** [review]
Thanks for doing this, just some minor suggestions and a question so I understand why we need to support the case where `O3DE_GEM_NAME` isn't defined.

File: AutomatedTesting/Gem/Code/Source/AutomatedTestingModule.cpp [hunk 4]
```diff
@@ -47,4 +47,8 @@ namespace AutomatedTesting
 // DO NOT MODIFY THIS LINE UNLESS YOU RENAME THE GEM
 // The first parameter should be GemName_GemIdLower
 // The second should be the fully qualified name of the class above
+#if defined(O3DE_GEM_NAME)
```
[the remark this answers is not in the document]
Comment (AMZN-alexpete) [reply AutomatedTesting/Gem/Code/Source/AutomatedTestingModule.cpp:50]: also, I think you can remove the comments on line 47-49

File: AutomatedTesting/Gem/CMakeLists.txt (hunk 1 above)
Comment (lumberyard-employee-dm) [reply AutomatedTesting/Gem/CMakeLists.txt:32]: That sounds good.

File: AutomatedTesting/Gem/Code/Source/AutomatedTestingModule.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (lumberyard-employee-dm) [reply AutomatedTesting/Gem/Code/Source/AutomatedTestingModule.cpp:50]: Well if it is not defined having a fallback of the original Gem name would still work as long as they user hasn't changed the name of the CMake Target for the game gem module.

File: Gems/AWSClientAuth/Code/CMakeLists.txt [hunk 5]
```diff
@@ -69,58 +78,58 @@ ly_add_target(
             AZ::AzToolsFramework
             Gem::AWSCore
             Gem::HttpRequestor
-            3rdParty::AWSNativeSDK::AWSClientAuth
+            3rdParty::AWSNativeSDK::${gem_name}
```
Comment (AMZN-Gene) [inline Gems/AWSClientAuth/Code/CMakeLists.txt:81]: I don't think we want to use $gem_name here. If this gem name changes, the aws sdk will still be called AWSClientAuth.

File: Gems/AWSClientAuth/Code/CMakeLists.txt [hunk 6]
```diff
@@ -143,19 +152,19 @@ if(PAL_TRAIT_BUILD_TESTS_SUPPORTED)
         BUILD_DEPENDENCIES
             PRIVATE
                 AZ::AzTest
-                3rdParty::AWSNativeSDK::AWSClientAuth
+                3rdParty::AWSNativeSDK::${gem_name}
```
Comment (AMZN-Gene) [inline Gems/AWSClientAuth/Code/CMakeLists.txt:155]: 3rd party, not our gem name

**Review by AMZN-Gene — APPROVED** [review]

File: Gems/AWSClientAuth/Code/CMakeLists.txt (hunk 5 above)
Comment (lumberyard-employee-dm) [reply Gems/AWSClientAuth/Code/CMakeLists.txt:81]: You are right, I careful not to change `3rdParty::AWSNativeSDK::AWSCore`, but I missed `3rdParty::AWSNativeSDK::AWSClientAuth`

File: Gems/Atom/Tools/MaterialCanvas/Code/CMakeLists.txt [hunk 7]
```diff
@@ -37,7 +37,7 @@ ly_add_target(
             Include
     BUILD_DEPENDENCIES
         PRIVATE
-            Gem::AtomLyIntegration_CommonFeatures.Static
+            Gem::CommonFeaturesAtom.Static
```
[the remark this answers is not in the document]
Comment (lumberyard-employee-dm) [reply Gems/Atom/Tools/MaterialCanvas/Code/CMakeLists.txt:40]: Yeah... both "Atom_Feature_Common" and "CommonFeaturesAtom" are terrible names for a Gem.
It doesn't hint at what kind of functionality those Gems could have on first read.

**Review by AMZN-alexpete — APPROVED** [review]
Amazing!  Thanks for making that macro, makes it so much more clean!

**Review by lsemp3d — APPROVED** [review]

File: Code/Framework/AzCore/AzCore/Memory/HphaAllocator.cpp [hunk 8]
```diff
@@ -1556,6 +1536,116 @@ namespace AZ
         return nullptr;
     }
 
+    template<bool DebugAllocatorEnable>
+    auto HphaSchemaBase<DebugAllocatorEnable>::HpAllocator::tree_grow_aligned(size_t size, size_t alignment) -> block_header*
+    {
+        const size_t sizeWithBlockHeaders = size + 3 * sizeof(block_header); // two fences plus one fake
+
+        // tree_system_alloc() will have an return an offset that is a multiple of OS_VIRTUAL_PAGE_SIZE which is 64KiB
```
Comment (spham-amzn) [inline Code/Framework/AzCore/AzCore/Memory/HphaAllocator.cpp:1544]:
```suggestion
        // tree_system_alloc() will return an offset that is a multiple of OS_VIRTUAL_PAGE_SIZE which is 64KiB
```

File: Code/Framework/AzCore/AzCore/Memory/HphaAllocator.cpp [hunk 9]
```diff
@@ -1556,6 +1536,116 @@ namespace AZ
         return nullptr;
     }
 
+    template<bool DebugAllocatorEnable>
+    auto HphaSchemaBase<DebugAllocatorEnable>::HpAllocator::tree_grow_aligned(size_t size, size_t alignment) -> block_header*
+    {
+        const size_t sizeWithBlockHeaders = size + 3 * sizeof(block_header); // two fences plus one fake
+
+        // tree_system_alloc() will return an offset that is a multiple of OS_VIRTUAL_PAGE_SIZE which is 64KiB
```
Comment (spham-amzn) [inline Code/Framework/AzCore/AzCore/Memory/HphaAllocator.cpp:1544]: I wonder if its better to put these longer explainations and details in a mark down file in this folder (something like HphaAllocator.md) and then just reference that file in the code instead.

Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/Memory/HphaAllocator.cpp:1544]: That could be arranged.

**Pull request merged**